shell/platform/fuchsia/dart_runner/dart_component_controller_v2.cc

```diff
@@ -0,0 +1,489 @@
+    FX_LOGF(ERROR, LOG_TAG, "Failed to FinalizeLoading: %s",
+            Dart_GetError(result));
+    Dart_ExitScope();
+    return false;
+  }
+
+  return true;
+}
+
+bool DartComponentControllerV2::SetUpFromAppSnapshot() {
+#if !defined(AOT_RUNTIME)
+  return false;
+#else
+  // Load the ELF snapshot as available, and fall back to a blobs snapshot
+  // otherwise.
+  const uint8_t *isolate_data, *isolate_instructions;
+  if (elf_snapshot_.Load(namespace_, data_path_ + "/app_aot_snapshot.so")) {
+    isolate_data = elf_snapshot_.IsolateData();
+    isolate_instructions = elf_snapshot_.IsolateInstrs();
+    if (isolate_data == nullptr || isolate_instructions == nullptr) {
+      return false;
+    }
+  } else {
+    if (!dart_utils::MappedResource::LoadFromNamespace(
+            namespace_, data_path_ + "/isolate_snapshot_data.bin",
+            isolate_snapshot_data_)) {
+      return false;
+    }
+    if (!dart_utils::MappedResource::LoadFromNamespace(
+            namespace_, data_path_ + "/isolate_snapshot_instructions.bin",
+            isolate_snapshot_instructions_, true /* executable */)) {
+      return false;
+    }
+  }
+  return CreateIsolate(isolate_data, isolate_instructions);
+#endif  // defined(AOT_RUNTIME)
+}
+
+bool DartComponentControllerV2::CreateIsolate(
+    const uint8_t* isolate_snapshot_data,
+    const uint8_t* isolate_snapshot_instructions) {
+  // Create the isolate from the snapshot.
+  char* error = nullptr;
+
+  // TODO(dart_runner): Pass if we start using tonic's loader.
+  intptr_t namespace_fd = -1;
+
+  // Freed in IsolateShutdownCallback.
+  auto state = new std::shared_ptr<tonic::DartState>(new tonic::DartState(
+      namespace_fd, [this](Dart_Handle result) { MessageEpilogue(result); }));
+
+  isolate_ = Dart_CreateIsolateGroup(
+      url_.c_str(), label_.c_str(), isolate_snapshot_data,
+      isolate_snapshot_instructions, nullptr /* flags */, state, state, &error);
+  if (!isolate_) {
+    FX_LOGF(ERROR, LOG_TAG, "Dart_CreateIsolateGroup failed: %s", error);
+    return false;
+  }
+
+  state->get()->SetIsolate(isolate_);
+
+  tonic::DartMessageHandler::TaskDispatcher dispatcher =
+      [loop = loop_.get()](auto callback) {
+        async::PostTask(loop->dispatcher(), std::move(callback));
+      };
+  state->get()->message_handler().Initialize(dispatcher);
+
+  state->get()->SetReturnCodeCallback(
+      [this](uint32_t return_code) { return_code_ = return_code; });
+
+  return true;
+}
+
+void DartComponentControllerV2::Run() {
+  async::PostTask(loop_->dispatcher(), [loop = loop_.get(), app = this] {
+    if (!app->RunDartMain()) {
+      loop->Quit();
+    }
+  });
+  loop_->Run();
+
+  if (binding_.is_bound()) {
+    binding_.Close(return_code_);
+  }
+}
+
+bool DartComponentControllerV2::RunDartMain() {
+  FML_CHECK(namespace_ != nullptr);
+  Dart_EnterScope();
+
+  tonic::DartMicrotaskQueue::StartForCurrentThread();
+
+  // TODO(fxb/...): Point these at the actual stdout/stderr handles of the
```

Comment:
There will never be actual stdout/stderr handles for the component. If we want to pass these in for each component then we will need to create them on our own and proxy them to the runner's stdout/stderr.

Reply:
I see. I think the distinction is a bit unclear to me - when you say "create them on our own and proxy them to the runner's stdout/stderr", do you mean the component creates them? And how does proxying them work again?

Maybe more generally is there more work required here? IIRC this was placeholder logic from the original PR.

Reply:
There is a bit of design work that needs to be done here so we can follow up more offline.

Fuchsia does not have a concept of stdout/stderr, what is happening is that the dart runner gets launched by the elf runner which creates a handle that can be used for stdout/stderr. Internally, it takes anything that is written to that descriptor and forwards it on to the kernel log and decorates that it is coming from the dart runner. If we pass along that same handle to the dart components then all print statements will appear to come from the dart runner. 

What I think we need to do is create a file descriptor for each component that is launched and listen for anything that is written to it. When something is written to it we just forward that message along to the fuchsia logger and decorate it with the tag that it came from the component.

Reply:
I see, makes sense. Thanks so much, updated the TODO here.